
## read into genomes, note currently the scripts only parse the sequence id with "Supercontig_12."
open GENOME, "<neurospora_crassa_or74a_12_supercontigs.fasta" or die "can not open genome file";
while (<GENOME>) { s/\r//g; chomp;
	if (/>(.+)/){
		$id = (split /\s+/, $1)[0];
	} else {
		$fas{$id} .= $_;
	}
}


## comment these lines if want to search directly without reverse complement
foreach $t ( keys %fas ) {
	$fas{$id} = reverse $fas{$id};
	$fas{$id} =~tr/ATCG/TAGC/;
}

## the length file contains the length of each chromosome and could be generated by "samtools faidx genome.fasta" 
open LEN, "<chr-length.txt" or die "Can not open genome file\.\n";
while (<LEN>) { s/\r//g; chomp; 
	@c = split /\t/;
	$len{$c[0]} = $c[1];
}


## read into all possible unit length combinations, e.g., 3^5 for 10, 11, 12 bp units
open ARRAY, "< unit_length_combinations.txt" or die "Can not open array file\.\n";
while (<ARRAY>) { chomp;
	@c = split /\t/; $group = join "\_", @c;    #$group1 = $c[0]; $i++; $iout = sprintf "%04d", $i;
	foreach $chr ( sort keys %fas ) {
		$seq = $fas{$chr};	$maxlen = $len{$chr};
#		warn "Initail task for chromosome $chr containing $maxlen bp\.\n";
		for ( $segstart=0; $segstart<= $maxlen-150; $segstart++ ) {
			$start=$segstart; 
			$new = substr($seq,$start,3);
			foreach $c (@c) {	$start+=$c; $new .= substr($seq,$start,3);  }
			$newn = substr($new,0,5);
			$hash{$newn} .= ">$chr\-$segstart\-$group\n$new\n" unless $new =~/N/;
		}
	}
	$ij++;  $time = localtime(); warn "$time processed array $ij $group\n";
}

## write the fragments to output files, compressed by gzip here
foreach $t ( keys %hash ) {
	open GROUP, " | gzip > f16\-minusrev\-u3g6\/mr\-$t\.fas\.gz" or die $!;
	print GROUP $hash{$t};
	close GROUP;
	delete $hash{$t};
}
